File: VS/Uses/InfixToPostfix.cpp
#include "C:\Users\Hussain\Desktop\DSA\VS\Classes\Stack\Stack.h"
#include <iostream>
using namespace std;

bool isValid(string);
bool isOperand(char);
bool isOperator(char);
bool isOpBracket(char);
bool isCloBracket(char);
int precedence(char);
bool priority(char, char);
void infixToPostfix();

//Main method
int main(){
    try{
        infixToPostfix();

    }catch(exception &e){
        cerr << e.what() << "\n\n";
    }
}

void infixToPostfix(){
    
    string s;
    cout << "\nEnter an Infix expression : "; getline(cin, s);
    string postfix;

    if(!isValid(s))
        throw runtime_error("Error! Invalid expression");

    Stack<char> stack(s.length());

    for(int i = 0; i < s.length(); i ++){
        
        char ch = s[i];
        
        if(ch == ' ') 
            continue;
        
        else if(isOpBracket(ch))
            stack.push(ch);

        else if(isOperand(ch))
            postfix += ch;
        
        else if(isOperator(ch)){
            if(stack.isEmpty())
                stack.push(ch);
            else{
                while(!stack.isEmpty() && priority(ch, stack.peek()))
                    postfix += stack.pop();
                stack.push(ch);
            }
        }
        else if(isCloBracket(ch)){

            switch (ch){
            case ')':
                while(!stack.isEmpty() && stack.peek() != '(')
                    postfix += stack.pop();
                break;
            case '}':
                while(!stack.isEmpty() && stack.peek() != '{')
                    postfix += stack.pop();
                break;
            case ']':
                while(!stack.isEmpty() && stack.peek() != '[')
                    postfix += stack.pop();
                break;
            }
            stack.pop();//rem opBracket from stack
        }
        
    }
    while(!stack.isEmpty())
        postfix += stack.pop();

    cout << "\nPostfix expression : " << postfix << "\n\n";
}

bool isValid(string s){
    Stack<char> stack(s.length());

    for(int i = 0; i < s.length(); i ++){
        
        char ch = s[i];

        if(isOpBracket(ch))
            stack.push(ch);
        
        else if(isCloBracket(ch)){
            if(stack.isEmpty())
                return false;

            else{
                switch (ch){
                case ')':
                    if(stack.pop() != '(')
                        return false;
                    break;
                case '}':
                    if(stack.pop() != '{')
                        return false;
                    break;
                case ']':
                    if(stack.pop() != '[')
                        return false;
                    break;
                }

            }
        }
        else if(isOperand(ch) || isOperator(ch))
            continue;

        else
            return false;
    }
    if(!stack.isEmpty())
        return false;
    
    return true;
}

bool isOperand(char ch){
    if(ch >= 'A' && ch <= 'Z' || ch >= 'a' && ch <= 'z' || ch >= '0' && ch <= '9')
        return true;
    return false;
}

bool isOperator(char ch){
    if(ch == '+' || ch == '-' || ch == '*' || ch == '/' || ch == '%' || ch == '^')
        return true;
    return false;
}

bool isOpBracket(char ch){
    if(ch == '(' || ch == '{' || ch == '[')
        return true;
    return false;
}

bool isCloBracket(char ch){
    if(ch == ')' || ch == '}' || ch == ']')
        return true;
    return false;
}

int precedence(char ch){
    switch(ch){

        case '^':
            return 3;
            break;

        case '*':
        case '/':
        case '%':
            return 2;
            break;

        case '+':
        case '-':
            return 1;
            break;
    }
}

bool priority(char ch, char stackTop){
    if(isOpBracket(stackTop) || precedence(ch) > precedence(stackTop))
        return false;//push
    if(precedence(ch) <= precedence(stackTop))
        return true;//check for next ch on stackTop
}

File: VS/Uses/InfixToPostfix.kt
import java.util.Scanner

fun main() {
    infixToPostfix()
}

fun infixToPostfix() {
    print("\nEnter an Infix expression : ")
    val input = Scanner(System.`in`)
    val str = input.nextLine()
    var postfix: String = ""

    if(!isValid(str)){
        println("\nError! Invalid expression\n")
        return
    }
    val stack = Stack<Char>()

    for(ch in str){
        if (ch == ' ')
            continue

        else if (isOpBracket(ch))
            stack.push(ch)

        else if (isOperand(ch))
            postfix += ch

        else if (isOperator(ch))
        {
            if (stack.isEmpty())
                stack.push(ch);
            else
            {
                while (!stack.isEmpty() && priority(ch, stack.peek()))
                    postfix += stack.pop();
                stack.push(ch);
            }
        }
        else if (isCloBracket(ch)){
            when(ch){
                ')' -> {
                    while (!stack.isEmpty() && stack.peek() != '(')
                        postfix += stack.pop()
                }
                '}' -> {
                    while (!stack.isEmpty() && stack.peek() != '{')
                        postfix += stack.pop()
                }
                ']' -> {
                    while (!stack.isEmpty() && stack.peek() != '[')
                        postfix += stack.pop()
                }
            }
            stack.pop(); // rem opBracket from stack
        }
    }
    while (!stack.isEmpty())
        postfix += stack.pop();

    print("\nPostfix expression : $postfix\n\n");
}

fun isValid(str: String): Boolean {
    val stack = Stack<Char>()
    for (ch in str) {
        if(isOpBracket(ch))
            stack.push(ch)

        else if(isCloBracket(ch)){
            if(stack.isEmpty())
                return false
            else{
                when(ch){
                    ')' -> {
                        if(stack.pop() != '(')
                            return false
                    }
                    '}' -> {
                        if (stack.pop() != '{')
                            return false;
                    }
                    ']' -> {
                        if (stack.pop() != '[')
                            return false;
                    }
                }
            }
        }
        else if (isOperand(ch) || isOperator(ch))
            continue;

        else
            return false;
    }
    if (!stack.isEmpty())
        return false;

    return true
}

fun isOperand(ch: Char) = ch in 'A'..'Z' || ch in 'a'..'z' || ch in '0'..'9'

fun isOperator(ch: Char) = when (ch) {
    '+', '-', '*', '/', '%', '^' -> true
        
    else -> false
}

fun isOpBracket(ch: Char) = when (ch) {
    '(', '{', '[' -> true

    else -> false
}

fun isCloBracket(ch: Char) = when (ch) {
    ')', '}', ']' -> true

    else -> false
}

fun precedence(ch: Char) = when(ch) {
    '^' -> 3

    '*', '/', '%' -> 2

    '+', '-' -> 1

    else -> -1
}

fun priority(ch: Char, stackTop: Char) = when {
    isOpBracket(stackTop) || precedence(ch) > precedence(stackTop) -> false; // push

    precedence(ch) <= precedence(stackTop) -> true; // check for next ch on stackTop

    else -> false
}

/* ******************************************************************************************* */

class Stack<T>(){

    private val items = mutableListOf<T>()

    fun isEmpty(): Boolean = if(items.isEmpty()) true else false

    fun size() = items.size

    fun peek() = items[items.size - 1]

    fun push(item: T){
        items.add(item)
        //println("Item pushed : $item")
    }

    fun pop(): T? {
        if(items.isEmpty()){
            println("Pop error! Stack is empty")
            return null
        }
        val temp = items.removeAt(items.size - 1)
        //println("Item popped : $temp")
        return temp
    }

    fun display() {
        if(items.isEmpty()){
            println("Error! Stack is empty")
            return
        }
        print("Stack -> ")
        for(i in items){
            print("$i ")
        }
        println()
    }
}